device/transport/amqp/src/amqp.ts

```diff
@@ -388,7 +613,11 @@ export class Amqp extends EventEmitter implements Client.Transport, StableConnec
    * @throws {ArgumentError}  One of the parameters is an incorrect type
    */
   sendTwinRequest(method: string, resource: string, properties: { [key: string]: string }, body: any, done?: (err?: Error, result?: results.MessageEnqueued) => void): void {
-    this._twinClient.sendTwinRequest(method, resource, properties, body, done);
+    this._fsm.handle('sendTwinRequest', method, resource, properties, body, (err) => {
+      if (done) {
+        done(err);
+      }
```

Comment:
<bert has an inner scream about consistency.  Why are some callbacks optional and others not?>

Reply:
yeah it's been a nagging pain for me too. we initially had very thin layers where almost all callbacks were optional. Since most of the API calls are internal i'm trying to push this only on the outer layer (in the Client) and make all other callbacks (internals) mandatory to make sure we explicitely handle all error cases. but for the sake of not breaking APIs, well.. sometimes I have to keep those.

I hope i'll be able to do a scrub of all this once we update the client.

---
In reply to: [141137658](https://github.com/Azure/azure-iot-sdk-node/pull/89#discussion_r141137658) [](ancestors = 141137658)